### PR TITLE
feat: add `cleanUrls` option (#219)

### LINKED ISSUE #219: Configurable 'clean url' in client router
**Is your feature request related to a problem? Please describe.**

In production it may be desirable to strip the `.html` extension in urls. Of course this isn't a direct concern of vitepress, and typically static hosts provide a turnkey setting to do this e.g; https://vercel.com/docs/configuration#project/clean-urls

However the client side router will append `.html` to the history push state when navigating:

- https://github.com/vuejs/vitepress/blob/master/src/client/app/router.ts#L47

This will result in an extra server redirects when e.g. sharing urls, doing hard reloads; that ideally could be avoided.

**Describe the solution you'd like**

To be able to configure this `.html` appending behaviour on the client router :)

### LINKED ISSUE #444: add meta to markdown-it env
### Is your feature request related to a problem? Please describe.

there is no context info currently, such as `filepath`/`vite compiler`, when `markdown-it` parse markdown files, and this leads to the inability to resolve relative file path when parsing markdown files, or add dependency to compiler to make hmr work as expected

for example, `snippet`

https://github.com/vuejs/vitepress/blob/ea6cb00c3e9099d45fc236bff0f7c42a65d558dc/src/node/markdown/plugins/snippet.ts?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L145-L147

```js
// loader always undefined
if (loader) {
  loader.addDependency(src)
}
```

### Describe the solution you'd like

add meta to [`markdown-it`](https://markdown-it.github.io/markdown-it/#MarkdownIt.render) be called) when [`render`](https://github.com/vuejs/vitepress/blob/ea6cb00c3e9099d45fc236bff0f7c42a65d558dc/src/node/markdown/markdown.ts?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L96-L105)

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://vuejs.org/coc)
- [X] Read the [docs](https://vitepress.vuejs.org/).
- [X] Read the [Contributing Guidelines](https://github.com/vuejs/vitepress/blob/master/.github/contributing.md).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by georges-gomes**
( Replaces #488 for alpha)

This feature is meant to be working with SPA and MPA mode so it's a little bit bigger than just rename the browser location.

- New option: cleanUrls (default false)
- Generates /foo.md -> /foo/index.html (This works best for MPA mode with most static site hosting services)
- Internal links in .md files are not transformed to .html
- When on SPA mode, the client router doesn't rename /foo into /foo.html anymore.
- Supports 404 page so it still generate `/404.html` (and not `/404/index.html` like other pages).

---

fixes #219, fixes #444

### CODE REVIEW

**Comment by brc-dd** [comment]
@kiaking I was thinking if instead of providing an option for `cleanUrls`, we can make it default? It'll simplify the code much more (like there will not be any need to pass it through different functions to `normalizeHref`). Also, this should fix our Algolia issue 😅.

**Comment by georges-gomes** [comment]
@brc-dd for people already using VitePress, this would change their URLs... It has a number of impacts for them. 
We can have `cleanUrls` = true for v1.0.0 (not my call) but I think we need this option for migration.

**Comment by georges-gomes** [comment]
@kiaking Does it looks good to you? Let me know if I should adjust something.

**Comment by azat-io** [comment]
Awesome 😍

File: src/node/config.ts [hunk 1]
```diff
@@ -71,6 +71,14 @@ export interface UserConfig<ThemeConfig = any> {
    * @default false
    */
   ignoreDeadLinks?: boolean
+
+  /**
+   * Always use "clean URLs" without the `.html`.
+   * Also generate static files as `foo/index.html` insted of `foo.html`.
+   * Works with MPA config too.
+   * (default: false)
```
Comment (brc-dd) [inline src/node/config.ts:79]:
```suggestion
   * @default false
```

File: src/node/markdown/markdown.ts [hunk 2]
```diff
@@ -56,7 +56,8 @@ export type { Header }
 export const createMarkdownRenderer = async (
   srcDir: string,
   options: MarkdownOptions = {},
-  base = '/'
+  base = '/',
+  cleanUrls: boolean = false
```
Comment (brc-dd) [inline src/node/markdown/markdown.ts:60]:
```suggestion
  cleanUrls = false
```

**Comment by brc-dd** [comment]
There is `.html` inside certain docs too. Since we're not automatically removing html from pages, it will be better to have consistent URLs in the docs. Also, it appears you've not allowed edits by maintainers in the PR.

**Comment by brc-dd** [comment]
This issue is also there: #412 if you add trailing slash to your URL. I had fixed this in vite-3 branch itself.

**Comment by brc-dd** [comment]
Also, there is some issue with sidebar. Open this: https://deploy-preview-869--vitepress-docs.netlify.app/guide/getting-started/, now click on configuration in sidebar. URL becomes this: `https://deploy-preview-869--vitepress-docs.netlify.app/guide/configuration.html`. Reload page it becomes: `https://deploy-preview-869--vitepress-docs.netlify.app/guide/configuration/`.

**Comment by georges-gomes** [comment]
> Also, there is some issue with sidebar. Open this: https://deploy-preview-869--vitepress-docs.netlify.app/guide/getting-started/, now click on configuration in sidebar. URL becomes this: `https://deploy-preview-869--vitepress-docs.netlify.app/guide/configuration.html`. Reload page it becomes: `https://deploy-preview-869--vitepress-docs.netlify.app/guide/configuration/`.

This is new, I have to check why. Thanks

**Comment by brc-dd** [comment]
You need to change this thing for sidebar issues IG:

https://github.com/vuejs/vitepress/pull/724/files#diff-4eb03faf823699119c94556187e2edb5771912371c55f9aef88804ec6c996564

**Comment by georges-gomes** [comment]
> You need to change this `normalizedPath` thing for sidebar issues IG:
> 
> https://github.com/vuejs/vitepress/pull/724/files#diff-4eb03faf823699119c94556187e2edb5771912371c55f9aef88804ec6c996564

Should be much better now.

**Comment by brc-dd** [comment]
> when you are happy with this change we can switch off the `cleanUrls` option before merging if you don't want it for the production site. But since we are bringing a brand new documentation, it would be the moment to bring this on... Your call.

Yeah, there is no need to disable it. 👍

**Comment by brc-dd** [comment]
Can you target this to `chore/vite-3` branch for now? Some fixes are not on `main` currently. That's why reloading pages is giving 404 with hydration mismatch.

File: src/client/theme-default/support/utils.ts [hunk 3]
```diff
@@ -69,17 +69,24 @@ export function normalize(path: string): string {
...
 
   const { pathname, search, hash } = new URL(url, 'http://example.com')
 
-  const normalizedPath =
-    pathname.endsWith('/') || pathname.endsWith('.html')
-      ? url
-      : `${pathname.replace(/(\.md)?$/, '.html')}${search}${hash}`
+  let normalizedPath = url
+  if (cleanUrls) {
+    normalizedPath = `${pathname.replace(/(\.md)?$/, '')}${search}${hash}`
```
Comment (brc-dd) [inline src/client/theme-default/support/utils.ts:81]:
```suggestion
    normalizedPath = `${pathname.replace(/(\.md)?$/, '/')}${search}${hash}`
```

File: src/client/app/router.ts [hunk 4]
```diff
@@ -42,9 +42,14 @@ export function createRouter(
...
-      url.pathname += '.html'
-      href = url.pathname + url.search + url.hash
+    // ensure correct deep link so page refresh lands on correct files.
+    if (siteDataRef.value.cleanUrls) {
+      // Should we replace `/foo.html` -> `/foo` ? :thinking:
+    } else {
+      if (!url.pathname.endsWith('/') && !url.pathname.endsWith('.html')) {
+        url.pathname += '.html'
+        href = url.pathname + url.search + url.hash
+      }
```
Comment (brc-dd) [inline src/client/app/router.ts:52]:
```suggestion
    if (!url.pathname.endsWith('/') && !url.pathname.endsWith('.html')) {
      url.pathname += siteDataRef.value.cleanUrls ? '/' : '.html'
      href = url.pathname + url.search + url.hash
```

**Comment by brc-dd** [comment]
@georges-gomes Without trailing slash Netlify is having issue. Open this site for example: https://62c2e731bab5b3761e2ce010--fastidious-stroopwafel-625e87.netlify.app/guide/getting-started

**Comment by georges-gomes** [comment]
> @georges-gomes Without trailing slash Netlify is having issue. Open this site for example: https://62c2e731bab5b3761e2ce010--fastidious-stroopwafel-625e87.netlify.app/guide/getting-started

Is Netlify setup with "Pretty URLs" ?

**Comment by brc-dd** [comment]
> Is Netlify setup with "Pretty URLs" ?

No asset optimization is disabled.

**Comment by brc-dd** [comment]
Okay, so this thing needs to be changed:

https://github.com/vuejs/vitepress/blob/37b89c07c953e8d1feea724ac9c4b6ce421c7af6/src/client/app/utils.ts#L25-L27

if clean url option is set we need to remove that `/` instead of adding `index`.

**Comment by brc-dd** [comment]
Having trailing slash will cause other issues too like a URL or asset like this `./theme-introduction` will resolve incorrectly. And Netlify apparently adds that:

> **With a file at /blog.html**
> - Navigating to /blog.html will serve /blog.html
> - Navigating to /blog will serve /blog
> - Navigating to /blog/ will 301 redirect and serve /blog
> 
> **With a file at /blog/index.html (and no file at /blog.html!)**
> - Navigating to /blog.html will 301 redirect and serve /blog/
> - Navigating to /blog will 301 redirect and serve /blog/
> - Navigating to /blog/ will serve /blog/

**Comment by georges-gomes** [comment]
When I use `npx serve` or Firebase hosting, I don't have issues. But we need to find the proper setup for Netlify I agree

**Comment by brc-dd** [comment]
The issue is on Vercel too, but in Vercel we can set trailingSlash redirects.

**Comment by georges-gomes** [comment]
For Netlify, this is not working? https://support.gatsbyjs.com/hc/en-us/articles/360056349113-Remove-trailing-slash-in-Netlify-URLs

**Comment by georges-gomes** [comment]
I'm reading that Netlify is biais toward trailing slash in our configuration. I think I'm going to option trailing and non-trailing slash option depending on the hosting solution. Nothing is easy.

**Comment by georges-gomes** [comment]
I created the option

## cleanUrls

- Type: `"off" | "with-trailing-slash" | "without-trailing-slash"`
- Default: `"off"`

When set to `"off"`, page `foo/bar.md` is generated into `foo/bar.html`.

When set to `"with-trailing-slash"` or `"without-trailing-slash"`, page `foo/bar.md` is generated into `foo/bar/index.html`.

When set to `"with-trailing-slash"`, URLs will be `foo/bar/`.
When set to `"without-trailing-slash"`, URLs will be `foo/bar`.

Notes:
- `404.md` page is kept transforming to `404.html` for hosting services.
- Also work in MPA mode.

**Comment by georges-gomes** [comment]
But there is still an issue with the router. It doesn't like trailing slashes. I couldn't find the solution quickly.

**Comment by georges-gomes** [comment]
OK trailing slash is fixed.

But there is a little challenge:

When `with-trailing-slash` is on, URL `foo/bar/` could be resolved to file `foo/bar.md` or `foo/bar/index.md`. For the moment I hardcoded the resolve  to `foo/bar.md`...

The `without-trailing-slash` option doesn't have this problem.

**Comment by brc-dd** [comment]
There some more issues with trailing slash. Like we have relative links in markdown. Click links in the "what's next" section on this page: http://localhost:4173/guide/getting-started/. They aren't being properly resolved. Also prefetching links is not working (see console for errors as you scroll to the bottom). I suppose the links can be handled in the markdown plugin. But there would be a challenge in determining if the page is `index` (no need to modify urls in this case) or if its just `foo` which we have made `foo/index` (will need to increase one level). Also this won't handle links inside Vue components.

**Comment by georges-gomes** [comment]
It's starting to require a lot of code to deal with trailing slash...
We better just not support it... and keep the option off while VitePress doc is hosted on Netlify...

**Comment by brc-dd** [comment]
I am thinking lets have three options:

1. Don't nest directories (no trailing slash) (just the router stuff, will work on almost all of the popular hosting services including Netlify, Vercel, GitHub Pages, Cloudflare Pages, cPanel/Apache/Nginx based hosting providers).

2. Nest directories (no trailing slash) will work with many hosting providers, especially the CDN type / static ones.

3. Off

**Comment by georges-gomes** [comment]
I don't understand what you call "nest directories" and "don't nest directory"

**Comment by brc-dd** [comment]
I meant this:

/foo.md -> /foo/index.html (create clean directory structure))
/foo.md -> /foo.html (don't create clean directory structure)

**Comment by georges-gomes** [comment]
I don't know if we can do (1) - it would rely of the hosting service to sert `/foo.html` when requested `/foo` without redirect... I can leave the door open for (1) but just implement (2) and (3) for now.

**Comment by brc-dd** [comment]
Fine!

**Comment by brc-dd** [comment]
Hi! Can you apply this diff: https://github.com/vuejs/vitepress/commit/e5856ca65d49fdc51b28e5fb3014c0dd60ea2bd6.diff (just download it to repo and `git apply filename.diff`).

**Comment by georges-gomes** [comment]
> Can you review #856 and verify if the clean urls are working cool with `with-subfolders` option? I had tested it on my setup. It worked fine. Just making sure twice.

Testing soon!

**Pull request closed without merging**